home/tests/test_display_question_feature.py

```diff
@@ -107,10 +91,97 @@ def test_response_context(self, response):
+            url = f'/explore/question_14/?sortanswersby={filter_type}'
             response = client.get(url)
-            answer = response.context['answers'].first()
+            answer = response.context['answers_structure'][0][0]
             assert str(answer.content) == expected
+
+    class TestLikesMechanism:
+        @pytest.mark.django_db
+        def test_logged_user_like(self, client):
+            """
+            in this test, a logged user press like on (answer A) which he disliked before
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "like 9"})
+            response = client.get(response.url)
+            answerA = '(<Answer: Answer A>, \'bi bi-hand-thumbs-up-fill\', \'bi bi-hand-thumbs-down'
+            assert answerA in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_dislike(self, client):
+            """
+            in this test, a logged user press dislike on (answer B) which he liked before
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "dislike 10"})
+            response = client.get(response.url)
+            answerB = '(<Answer: Answer B>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down-fill'
+            assert answerB in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_unlike(self, client):
+            """
+            in this test, a logged user press like on (answer B) which he liked before
+            and make it un-liked
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "like 10"})
+            response = client.get(response.url)
+            answerB = '(<Answer: Answer B>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down'
+            assert answerB in str(response.context['answers_structure'])
+
+        @pytest.mark.django_db
+        def test_logged_user_undislike(self, client):
+            """
+            in this test, a logged user press dislike on (answer A) which he disliked before
+            and make it un-disliked
+            """
+            user = Profile.objects.all().get(user_id=1).user
+            client.force_login(user)
+            url = reverse('thumbs')
+            response = client.post(url, {'button': "dislike 9"})
+            response = client.get(response.url)
+            answerA = '(<Answer: Answer A>, \'bi bi-hand-thumbs-up\', \'bi bi-hand-thumbs-down'
+            assert answerA in str(response.context['answers_structure'])
```

Comment:
How do you know what were the choices of the user before in these 4 methods?

Reply:
By the "answer_test_data" variable of 0015_add_answers_test_data.py file